cpp/include/cudf/detail/utilities/cuda_memcpy.hpp

```diff
@@ -17,41 +17,97 @@
 #pragma once
 
 #include <cudf/utilities/export.hpp>
+#include <cudf/utilities/span.hpp>
 
 #include <rmm/cuda_stream_view.hpp>
 
 namespace CUDF_EXPORT cudf {
 namespace detail {
 
+namespace impl {
```

Comment:
Why do we need an `impl` inside of `detail`? Isn't `detail` internal-enough?

Reply:
ah, just realized!
I should just name the function `cuda_memcpy_async_impl` and drop the `impl` namespace, right?

Reply:
My guess is that we'd like for consumers to use `cuda_memcpy_async<T>` instead of the raw `cuda_memcpy_async(void*, void*)`.

Ideally, this would have been:
1. hidden away in a separate `.cu` (but it's used from a function template), 
2. Or in the unnamed namespace (which hides it not at all in an `.hpp`).

I'm curious:  What _is_ the recommended way to hide an impl function that's used from a function template?

Reply:
(Ah, I hadn't seen @vuule's reply before posting.)